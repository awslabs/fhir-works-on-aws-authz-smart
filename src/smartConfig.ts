/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */
import { KeyValueMap } from 'fhir-works-on-aws-interface';

export type ScopeType = 'patient' | 'user';
export type AccessModifier = 'read' | 'write' | '*';
export type IdentityType = 'Patient' | 'Practitioner' | 'Person ' | 'RelatedPerson';

export interface ClinicalSmartScope {
    scopeType: ScopeType;
    resourceType: string;
    accessType: AccessModifier;
}

export type AccessRule = {
    read: (
        | 'read'
        | 'vread'
        | 'history-type'
        | 'history-instance'
        | 'search-type'
        | 'transaction'
        | 'batch'
        | 'search-system'
        | 'history-system'
    )[];
    write: ('transaction' | 'batch' | 'create' | 'update' | 'delete' | 'patch')[];
};

/**
 * Determines what each scope has access to do
 * Scope `patient/Patient.read` maps to `scopeRule.patient.read` operations
 *  @example
 * {
 *      patient: {
 *          read: ['read','search-type'],
 *          write: [],
 *      },
 *      user: {
 *          read: ['read','search-type', 'vread'],
 *          write: ['transaction','update', 'patch', 'create'],
 *      },
 *  };
 */
export interface ScopeRule {
    patient: AccessRule;
    user: AccessRule;
}

export type FhirResource = { hostname: string; resourceType: string; id: string };

export interface UserIdentity extends KeyValueMap {
    scopes: string[];
    fhirUserObject?: FhirResource;
    patientLaunchContext?: FhirResource;
}

export interface SMARTConfig {
    version: number;
    /**
     * Within the access_token the scopes are typically sent in the 'scp' or 'scope' key
     */
    scopeKey: string;
    /**
     * Defined more below
     */
    scopeRule: ScopeRule;
    /**
     * Per SMART spec this is the 'aud' key found in the access_token
     */
    expectedAudValue: string;
    /**
     * Per SMART spec this is the 'iss' key found in the access_token
     */
    expectedIssValue: string;
    /**
     * Name of the claim found in the access_token that represents the requestors FHIR Id
     */
    fhirUserClaimKey: 'fhirUser' | 'profile';
    /**
     * Prefix of the claim found in the access_token that represents the requestors launch context. The remaining part of the claim will identify the resource type i.e. `launch_response_patient`
     * @example launch_response_
     */
    launchContextKeyPrefix: string;
    /**
     * Json Web Key Set endpoint used to get the key for verifying access_token
     */
    jwksEndpoint: string;
}
