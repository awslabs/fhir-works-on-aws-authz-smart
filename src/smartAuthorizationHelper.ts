/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */
import { FhirVersion, UnauthorizedError } from 'fhir-works-on-aws-interface';
import jwksClient, { JwksClient } from 'jwks-rsa';
import { decode, verify } from 'jsonwebtoken';
import resourceReferencesMatrixV4 from './schema/fhirResourceReferencesMatrix.v4.0.1.json';
import resourceReferencesMatrixV3 from './schema/fhirResourceReferencesMatrix.v3.0.1.json';
import { FhirResource } from './smartConfig';

export const FHIR_USER_REGEX = /^(?<hostname>(http|https):\/\/([A-Za-z0-9\-\\.:%$_/])+)\/(?<resourceType>Person|Practitioner|RelatedPerson|Patient)\/(?<id>[A-Za-z0-9\-.]+)$/;
export const FHIR_RESOURCE_REGEX = /^((?<hostname>(http|https):\/\/([A-Za-z0-9\-\\.:%$_/])+)\/)?(?<resourceType>[A-Z][a-zA-Z]+)\/(?<id>[A-Za-z0-9\-.]+)$/;

export function getFhirUser(fhirUserValue: string): FhirResource {
    const match = fhirUserValue.match(FHIR_USER_REGEX);
    if (match) {
        const { hostname, resourceType, id } = match.groups!;
        return { hostname, resourceType, id };
    }
    throw new UnauthorizedError("Requester's identity is in the incorrect format");
}
export function getFhirResource(resourceValue: string, defaultHostname: string): FhirResource {
    const match = resourceValue.match(FHIR_RESOURCE_REGEX);
    if (match) {
        const { resourceType, id } = match.groups!;
        const hostname = match.groups!.hostname ?? defaultHostname;
        return { hostname, resourceType, id };
    }
    throw new UnauthorizedError('Resource is in the incorrect format');
}

function isRequestorReferenced(
    requestorIds: string[],
    requestorResourceType: string,
    sourceResource: any,
    fhirVersion: FhirVersion,
): boolean {
    const sourceResourceType = sourceResource.resourceType;
    let matrix: any;
    if (fhirVersion === '4.0.1') {
        matrix = resourceReferencesMatrixV4;
    } else if (fhirVersion === '3.0.1') {
        matrix = resourceReferencesMatrixV3;
    } else {
        throw new Error('Unsupported FHIR version detected');
    }
    let possiblePaths: string[] = [];
    if (matrix[sourceResourceType] && matrix[sourceResourceType][requestorResourceType]) {
        possiblePaths = matrix[sourceResourceType][requestorResourceType];
    }

    // The paths within the FHIR resources may contain arrays so we must check if array at every level
    return possiblePaths.some(path => {
        const pathComponents: string[] = path.split('.');
        let tempResource = sourceResource;
        let rootQueue = [];
        let nextQueue = [tempResource[pathComponents[0]]];
        for (let i = 1; i < pathComponents.length; i += 1) {
            rootQueue = nextQueue;
            nextQueue = [];

            while (rootQueue.length > 0) {
                tempResource = rootQueue.shift();
                if (tempResource) {
                    if (Array.isArray(tempResource)) {
                        // eslint-disable-next-line no-loop-func
                        tempResource.forEach(x => {
                            nextQueue.push(x[pathComponents[i]]);
                        });
                    } else {
                        nextQueue.push(tempResource[pathComponents[i]]);
                    }
                }
            }
        }
        return nextQueue.flat().some(x => {
            return x && x.reference && requestorIds.includes(x.reference);
        });
    });
}

export function hasReferenceToResource(
    requestorId: FhirResource,
    sourceResource: any,
    apiUrl: string,
    fhirVersion: FhirVersion,
): boolean {
    const { hostname, resourceType, id } = requestorId;
    if (hostname !== apiUrl) {
        // If requester is not from this FHIR Server they must be a fully qualified reference
        return isRequestorReferenced([`${hostname}/${resourceType}/${id}`], resourceType, sourceResource, fhirVersion);
    }

    return (
        (resourceType === sourceResource.resourceType && id === sourceResource.id) ||
        isRequestorReferenced(
            [`${resourceType}/${id}`, `${hostname}/${resourceType}/${id}`],
            resourceType,
            sourceResource,
            fhirVersion,
        )
    );
}

export function getJwksClient(jwksUri: string): JwksClient {
    return jwksClient({
        cache: true,
        cacheMaxEntries: 5,
        cacheMaxAge: 600000,
        rateLimit: true,
        jwksRequestsPerMinute: 10,
        jwksUri,
    });
}

export async function verifyJwtToken(
    token: string,
    expectedAudValue: string,
    expectedIssValue: string,
    client: JwksClient,
) {
    const genericErrorMessage = 'Invalid access token';
    const decodedAccessToken = decode(token, { complete: true });
    if (decodedAccessToken === null || typeof decodedAccessToken === 'string') {
        console.error('access_token could not be decoded into an object');
        throw new UnauthorizedError(genericErrorMessage);
    }
    const { kid } = decodedAccessToken.header;
    if (!kid) {
        console.error('JWT verification failed. JWT "kid" attribute is required in the header');
        throw new UnauthorizedError(genericErrorMessage);
    }

    try {
        const key = await client.getSigningKeyAsync(kid);
        return verify(token, key.getPublicKey(), { audience: expectedAudValue, issuer: expectedIssValue });
    } catch (e) {
        console.error(e.message);
        throw new UnauthorizedError(genericErrorMessage);
    }
}
