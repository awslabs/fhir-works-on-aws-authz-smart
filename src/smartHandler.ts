/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */
import { decode } from 'jsonwebtoken';
import {
    Authorization,
    VerifyAccessTokenRequest,
    AuthorizationBundleRequest,
    AllowedResourceTypesForOperationRequest,
    SystemOperation,
    TypeOperation,
    UnauthorizedError,
    ReadResponseAuthorizedRequest,
    WriteRequestAuthorizedRequest,
    AccessBulkDataJobRequest,
    R4_PATIENT_COMPARTMENT_RESOURCES,
    KeyValueMap,
} from 'fhir-works-on-aws-interface';
import axios from 'axios';
import { LaunchType, ScopeType, SMARTConfig } from './smartConfig';

// eslint-disable-next-line import/prefer-default-export
export class SMARTHandler implements Authorization {
    static readonly CLINICAL_SCOPE_REGEX = /^(?<scopeType>patient|user|system)\/(?<scopeResourceType>[a-zA-Z]+|\*)\.(?<accessType>read|write|\*)$/;

    static readonly LAUNCH_SCOPE_REGEX = /^(?<scopeType>launch)(\/(?<launchType>patient|encounter))?$/;

    private readonly version: number = 1.0;

    private readonly config: SMARTConfig;

    constructor(config: SMARTConfig) {
        if (config.version !== this.version) {
            throw Error('Authorization configuration version does not match handler version');
        }
        this.config = config;
    }

    async verifyAccessToken(request: VerifyAccessTokenRequest): Promise<KeyValueMap> {
        if (
            request.operation === 'read' &&
            (request.resourceType === 'metadata' || request.resourceType === '.well-known')
        ) {
            return {};
        }

        // The access_token will be verified by hitting the authZUserInfoUrl (token introspection)
        // Decoding first to determine if it passes scope & claims check first
        const decoded = decode(request.accessToken, { json: true }) || {};
        const { aud, iss } = decoded;
        const audArray = Array.isArray(aud) ? aud : [aud];

        // verify aud & iss
        if (!audArray.includes(this.config.expectedAudValue) || this.config.expectedIssValue !== iss) {
            console.error('aud or iss is not matching');
            throw new UnauthorizedError('Error validating the validity of the access_token');
        }

        // verify scope
        let scopes: string[] = [];
        if (this.config.scopeValueType === 'space' && typeof decoded[this.config.scopeKey] === 'string') {
            scopes = decoded[this.config.scopeKey].split(' ');
        } else if (this.config.scopeValueType === 'array' && Array.isArray(decoded[this.config.scopeKey])) {
            scopes = decoded[this.config.scopeKey];
        }
        if (!this.areScopesSufficient(scopes, request.operation, request.resourceType)) {
            console.error(
                `User supplied scopes are insufficient\nscopes: ${scopes}\noperation: ${request.operation}\nresourceType: ${request.resourceType}`,
            );
            throw new UnauthorizedError('User does not have permission for requested operation');
        }

        // Verify token
        let response;
        try {
            response = await axios.get(this.config.userInfoEndpoint, {
                headers: { Authorization: `Bearer ${request.accessToken}` },
            });
        } catch (e) {
            console.error('Post to authZUserInfoUrl failed', e);
        }
        if (!response || !response.data[this.config.expectedFhirUserClaimKey]) {
            console.error(`result from AuthZ did not have ${this.config.expectedFhirUserClaimKey} claim`);
            throw new UnauthorizedError("Cannot determine requester's identity");
        }
        return response.data;
    }

    // eslint-disable-next-line class-methods-use-this, @typescript-eslint/no-unused-vars
    async isAccessBulkDataJobAllowed(request: AccessBulkDataJobRequest): Promise<void> {
        // TODO not supported for now
        throw new UnauthorizedError('Bulk access is not allowed');
    }

    // eslint-disable-next-line class-methods-use-this, @typescript-eslint/no-unused-vars
    async isBundleRequestAuthorized(request: AuthorizationBundleRequest): Promise<void> {
        // TODO not supported for now
        throw new UnauthorizedError('Bundle operation is not authorzied');
    }

    // eslint-disable-next-line class-methods-use-this, @typescript-eslint/no-unused-vars
    async getAllowedResourceTypesForOperation(request: AllowedResourceTypesForOperationRequest): Promise<string[]> {
        // TODO this is stubbed for now
        return R4_PATIENT_COMPARTMENT_RESOURCES;
    }

    // eslint-disable-next-line class-methods-use-this, @typescript-eslint/no-unused-vars
    async authorizeAndFilterReadResponse(request: ReadResponseAuthorizedRequest): Promise<any> {
        // TODO this is stubbed for now
        return request.readResponse;
    }

    // eslint-disable-next-line class-methods-use-this, @typescript-eslint/no-unused-vars
    async isWriteRequestAuthorized(request: WriteRequestAuthorizedRequest): Promise<void> {
        // TODO this is stubbed for now
    }

    private areScopesSufficient(
        scopes: string[],
        operation: TypeOperation | SystemOperation,
        resourceType?: string,
    ): boolean {
        const { scopeRule } = this.config;
        for (let i = 0; i < scopes.length; i += 1) {
            const scope = scopes[i];
            let match = scope.match(SMARTHandler.LAUNCH_SCOPE_REGEX);
            if (!match) {
                match = scope.match(SMARTHandler.CLINICAL_SCOPE_REGEX);
            }
            if (match !== null) {
                const { scopeType } = match.groups!;
                let validOperations: (TypeOperation | SystemOperation)[] = [];
                if (scopeType === 'launch') {
                    const { launchType } = match.groups!;
                    // TODO: should launch have access to only certain resourceTypes?
                    if (['patient', 'encounter'].includes(launchType)) {
                        validOperations = scopeRule[scopeType][<LaunchType>launchType];
                    } else if (!launchType) {
                        validOperations = scopeRule[scopeType].launch;
                    }
                } else if (['patient', 'user', 'system'].includes(scopeType)) {
                    const { scopeResourceType, accessType } = match.groups!;
                    if (resourceType) {
                        if (scopeResourceType === '*' || scopeResourceType === resourceType) {
                            validOperations = this.getValidOperationsForScope(<ScopeType>scopeType, accessType);
                        }
                    } else if (scopeResourceType === '*') {
                        validOperations = this.getValidOperationsForScope(<ScopeType>scopeType, accessType);
                    }
                }
                const isAuthorized = validOperations.includes(operation);
                if (isAuthorized) return true;
            }
        }
        return false;
    }

    private getValidOperationsForScope(scopeType: ScopeType, accessType: string) {
        let validOperations: (TypeOperation | SystemOperation)[] = [];
        if (accessType === '*' || accessType === 'read') {
            validOperations = this.config.scopeRule[scopeType].read;
        }
        if (accessType === '*' || accessType === 'write') {
            validOperations = validOperations.concat(this.config.scopeRule[scopeType].write);
        }
        return validOperations;
    }
}
