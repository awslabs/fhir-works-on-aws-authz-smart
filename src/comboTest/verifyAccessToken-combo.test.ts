// import { BulkDataAuth, SystemOperation, TypeOperation } from 'fhir-works-on-aws-interface';
// import * as fs from 'fs';
// import { filterOutUnusableScope } from './smartScopeHelper';
import * as path from 'path';
import { VerifyAccessTokenRequest } from 'fhir-works-on-aws-interface';
import { ScopeRule, SMARTConfig } from '../smartConfig';
import { SMARTHandler } from '../smartHandler';
import * as smartAuthorizationHelper from '../smartAuthorizationHelper';

// const sync = require('csv-parse/lib/sync');

const { load } = require('csv-load-sync');

const scopeRule = (): ScopeRule => ({
    patient: {
        read: ['read', 'vread', 'search-type', 'search-system', 'history-instance', 'history-type', 'history-system'],
        write: ['create', 'transaction'],
    },
    user: {
        read: ['read', 'vread', 'search-type', 'search-system', 'history-instance', 'history-type', 'history-system'],
        write: ['update', 'patch', 'create', 'delete', 'transaction'],
    },
    system: {
        // "read" allows system export and group export
        read: ['read'],
        write: [],
    },
});

interface CsvRow {
    patientContext: string;
    fhirUser: string;
    operation: string;
    resourceType: string;
    id: string;
    vid: string;
    fhirServiceBaseUrl: string;
    'patient/Patient.read': string;
    'patient/Patient.write': string;
    'patient/MedicationRequest.read': string;
    'patient/MedicationRequest.write': string;
    'patient/Binary.read': string;
    'patient/Binary.write': string;
    'user/Patient.read': string;
    'user/Patient.write': string;
    'user/MedicationRequest.read': string;
    'user/MedicationRequest.write': string;
    'user/Binary.read: string': string;
    'user/Binary.write': string;
    'system/Patient.read': string;
    'system/Patient.write': string;
    'system/MedicationRequest.read': string;
    'system/MedicationRequest.write': string;
    'system/Binary.read': string;
    'system/Binary.write': string;
}

const getScopesFromResult = (result: any) => {
    const scopes: string[] = [];
    Object.entries(result).forEach(([key, value]) => {
        if (key.startsWith('system/') || key.startsWith('user/') || (key.startsWith('/patient/') && value === key)) {
            scopes.push(key);
        }
    });
    return scopes;
};
// TODO : use path to get the right path
const csv: CsvRow[] = load(path.resolve(__dirname, './params/VerifyAccessToken-NoBulkDataAuth-params.csv'), {
    convert: {
        isUserScopeAllowedForSystemExport: (s: string) => s === 'true',
    },
});
const clonedScopeRule = scopeRule();
const testCasesFromCSV: any[] = [];
csv.forEach((row) => {
    const testInput: any = {
        scopes: getScopesFromResult(row),
        scopeRule: clonedScopeRule,
        // reqOperation: row.operation as TypeOperation | SystemOperation,
        // isUserScopeAllowedForSystemExport: row.isUserScopeAllowedForSystemExport,
        // reqResourceType: row.reqResourceType,
        bulkDataAuth: undefined,
        patientContext: row.patientContext,
        fhirUser: row.fhirUser,
    };
    testCasesFromCSV.push(testInput);
});

// TODO: Combine common set up to a separate util file
const expectedAud = 'api://default';
const expectedIss = 'https://dev-6460611.okta.com/oauth2/default';
const baseAuthZConfig = (): SMARTConfig => ({
    version: 1.0,
    scopeKey: 'scp',
    scopeRule: scopeRule(),
    expectedAudValue: expectedAud,
    expectedIssValue: expectedIss,
    fhirUserClaimPath: 'fhirUser',
    launchContextPathPrefix: 'ext.launch_response_',
    jwksEndpoint: `${expectedIss}/jwks`,
});
const apiUrl = 'https://fhir.server.com/dev';
const id = 'id';
const patientId = `Patient/${id}`;
const practitionerId = `Practitioner/${id}`;
const patientIdentity = `${apiUrl}/${patientId}`;
const practitionerIdentity = `${apiUrl}/${practitionerId}`;
const sub = 'test@test.com';

const patientContext: any = {
    ext: { launch_response_patient: patientIdentity },
};
const patientFhirUser: any = {
    fhirUser: patientIdentity,
};
const practitionerFhirUser: any = {
    fhirUser: practitionerIdentity,
};

const baseAccessNoScopes: any = {
    ver: 1,
    jti: 'AT.6a7kncTCpu1X9eo2QhH1z_WLUK4TyV43n_9I6kZNwPY',
    iss: expectedIss,
    aud: expectedAud,
    iat: Math.floor(Date.now() / 1000) - 1,
    exp: Math.floor(Date.now() / 1000) + 60 * 60,
    cid: '0oa8muazKSyk9gP5y5d5',
    uid: '00u85ozwjjWRd17PB5d5',
    sub,
};
// End TODO

describe('verifyAccessToken-combo', () => {
    // Example array of test cases
    // TODO: next step is to convert csv file loaded to follow the same format
    const cases: (string | boolean | VerifyAccessTokenRequest | any)[][] = [
        [
            'Read and Write Access: initiate-export',
            {
                accessToken: 'fake',
                operation: 'read',
                resourceType: '',
                bulkDataAuth: { exportType: 'system', operation: 'initiate-export' },
            },
            { ...baseAccessNoScopes, scp: ['user/*.*', 'patient/*.write'], ...practitionerFhirUser },
        ],
        [
            'Read and Write Access: initiate-export',
            {
                accessToken: 'fake',
                operation: 'read',
                resourceType: '',
                bulkDataAuth: { exportType: 'system', operation: 'initiate-export' },
            },
            { ...baseAccessNoScopes, scp: ['user/*.*', 'patient/*.write'], ...patientFhirUser },
        ],
        [
            'Read and Write Access: initiate-export',
            {
                accessToken: 'fake',
                operation: 'read',
                resourceType: '',
                bulkDataAuth: { exportType: 'system', operation: 'initiate-export' },
            },
            { ...baseAccessNoScopes, scp: ['user/*.*', 'patient/*.write'], ...patientContext },
        ],
    ];

    const authZConfig = baseAuthZConfig();
    const authZHandler: SMARTHandler = new SMARTHandler(authZConfig, apiUrl, '4.0.1', undefined, undefined, true);
    test.each(cases)('CASE: %p', async (_firstArg, request, decodedAccessToken) => {
        const authZHandlerWithAnotherApiURL: SMARTHandler = new SMARTHandler(
            authZConfig,
            'https://some-server.com',
            '4.0.1',
            undefined,
            undefined,
            true,
        );

        const requestWithFhirServiceBaseUrl = { ...request, fhirServiceBaseUrl: apiUrl };

        // Handling mocking modules when code is in TS: https://stackoverflow.com/a/60693903/14310364
        jest.spyOn(smartAuthorizationHelper, 'verifyJwtToken').mockImplementation(() =>
            Promise.resolve(decodedAccessToken),
        );

        // TODO: Snapshot contains timestamp, need to update logic to static or it fails on rerun
        try {
            await expect(authZHandler.verifyAccessToken(<VerifyAccessTokenRequest>request)).resolves.toMatchSnapshot();
        } catch (e) {
            expect(e).toMatchSnapshot();
        }

        try {
            await expect(
                authZHandlerWithAnotherApiURL.verifyAccessToken(
                    <VerifyAccessTokenRequest>requestWithFhirServiceBaseUrl,
                ),
            ).resolves.toMatchSnapshot();
        } catch (e) {
            expect(e).toMatchSnapshot();
        }
    });
});
