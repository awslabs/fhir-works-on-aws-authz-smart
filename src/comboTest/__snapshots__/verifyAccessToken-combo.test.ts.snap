// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verifyAccessToken-combo CASE: "Read and Write Access: initiate-export" 1`] = `
Object {
  "aud": "api://default",
  "cid": "0oa8muazKSyk9gP5y5d5",
  "exp": 1668482781,
  "fhirUser": "https://fhir.server.com/dev/Practitioner/id",
  "fhirUserObject": Object {
    "hostname": "https://fhir.server.com/dev",
    "id": "id",
    "resourceType": "Practitioner",
  },
  "iat": 1668479180,
  "iss": "https://dev-6460611.okta.com/oauth2/default",
  "jti": "AT.6a7kncTCpu1X9eo2QhH1z_WLUK4TyV43n_9I6kZNwPY",
  "scopes": Array [
    "user/*.*",
    "patient/*.write",
  ],
  "scp": Array [
    "user/*.*",
    "patient/*.write",
  ],
  "sub": "test@test.com",
  "uid": "00u85ozwjjWRd17PB5d5",
  "usableScopes": Array [
    "user/*.*",
  ],
  "ver": 1,
}
`;

exports[`verifyAccessToken-combo CASE: "Read and Write Access: initiate-export" 2`] = `
Object {
  "aud": "api://default",
  "cid": "0oa8muazKSyk9gP5y5d5",
  "exp": 1668482781,
  "fhirUser": "https://fhir.server.com/dev/Practitioner/id",
  "fhirUserObject": Object {
    "hostname": "https://fhir.server.com/dev",
    "id": "id",
    "resourceType": "Practitioner",
  },
  "iat": 1668479180,
  "iss": "https://dev-6460611.okta.com/oauth2/default",
  "jti": "AT.6a7kncTCpu1X9eo2QhH1z_WLUK4TyV43n_9I6kZNwPY",
  "scopes": Array [
    "user/*.*",
    "patient/*.write",
  ],
  "scp": Array [
    "user/*.*",
    "patient/*.write",
  ],
  "sub": "test@test.com",
  "uid": "00u85ozwjjWRd17PB5d5",
  "usableScopes": Array [
    "user/*.*",
  ],
  "ver": 1,
}
`;

exports[`verifyAccessToken-combo CASE: "Read and Write Access: initiate-export" 3`] = `
[Error: [2mexpect([22m[31mreceived[39m[2m).[22mresolves[2m.[22mtoMatchSnapshot[2m()[22m

Received promise rejected instead of resolved
Rejected to value: [31m[UnauthorizedError: User does not have permission for requested operation][39m]
`;

exports[`verifyAccessToken-combo CASE: "Read and Write Access: initiate-export" 4`] = `
[Error: [2mexpect([22m[31mreceived[39m[2m).[22mresolves[2m.[22mtoMatchSnapshot[2m()[22m

Received promise rejected instead of resolved
Rejected to value: [31m[UnauthorizedError: User does not have permission for requested operation][39m]
`;

exports[`verifyAccessToken-combo CASE: "Read and Write Access: initiate-export" 5`] = `
[Error: [2mexpect([22m[31mreceived[39m[2m).[22mresolves[2m.[22mtoMatchSnapshot[2m()[22m

Received promise rejected instead of resolved
Rejected to value: [31m[Error: No usable scope][39m]
`;

exports[`verifyAccessToken-combo CASE: "Read and Write Access: initiate-export" 6`] = `
[Error: [2mexpect([22m[31mreceived[39m[2m).[22mresolves[2m.[22mtoMatchSnapshot[2m()[22m

Received promise rejected instead of resolved
Rejected to value: [31m[Error: No usable scope][39m]
`;
